File: Binary Search/3488. Closest Equal Element Queries/solution.py
class Solution:
    def solveQueries(self, nums: List[int], queries: List[int]) -> List[int]:
        n = len(nums)

        indice = defaultdict(list)
        for i, num in enumerate(nums):
            indice[num].append(i)


        res = []
        for i in queries:
            if (m := len(indice[nums[i]])) > 1:
                r = bisect_right(indice[nums[i]], i)%m # right closest, take modular to find circular item if not exist
                l = (bisect_left(indice[nums[i]], i)-1+m)%m # left closest, take modular to find circular item if not exist
                idxR = indice[nums[i]][r]
                idxL = indice[nums[i]][l]
                dist1 = abs(i-idxR) # n-dist1 = ther other circular distance
                dist2 = abs(i-idxL) # n-dist2 = ther other circular distance
                res.append(min(min(dist1, n-dist1), min(dist2, n-dist2)))
            else:
                res.append(-1)
        return res

File: Binary Search/3488. Closest Equal Element Queries/solution.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<int> solveQueries(vector<int>& nums, vector<int>& queries) {
        int n = nums.size();
        unordered_map<int, vector<int>> valToIndex;
        
        // Step 1: Group indices by their values
        for (int i = 0; i < n; i++) {
            valToIndex[nums[i]].push_back(i);
        }

        vector<int> res;
        for (auto& targetIdx : queries) {
            const vector<int>& indices = valToIndex[nums[targetIdx]];

            // If only one such element exists, there's no "other" index j
            if (indices.size() <= 1) {
                res.push_back(-1);
                continue;
            }

            // Step 2: Use binary search to find the position of targetIdx in the index list
            auto it = lower_bound(indices.begin(), indices.end(), targetIdx);
            int pos = distance(indices.begin(), it);
            int m = indices.size();

            // Step 3: Check neighbors (with circular wrapping)
            // The closest element must be the one immediately before or after in the sorted list
            int prevIdx = indices[(pos - 1 + m) % m];
            int nextIdx = indices[(pos + 1) % m];

            // Step 4: Calculate circular distances
            auto getDist = [&](int i, int j) {
                int d = abs(i - j);
                return min(d, n - d);
            };

            res.push_back(min(getDist(targetIdx, prevIdx), getDist(targetIdx, nextIdx)));
        }

        return res;
    }
};